### PR TITLE
Adds function for enabling tls client authentication

### PR DESCRIPTION
**Pull request by bkchr**
Requires: https://github.com/h2o/picotls/pull/121

Sorry for spamming you with pull requests that depend on other pull requests ^^

### CODE REVIEW

File: picoquic/picoquic.h [hunk 1]
```diff
@@ -210,6 +210,9 @@ void picoquic_free(picoquic_quic_t* quic);
 /* Set cookie mode on QUIC context when under stress */
 void picoquic_set_cookie_mode(picoquic_quic_t* quic, int cookie_mode);
 
+/* Enable the client authentication in TLS (client is required to send certificates). */
+void picoquic_enable_client_authentication(picoquic_quic_t* quic);
+
```
Comment (huitema) [inline picoquic/picoquic.h]: How does an application disable this option if it is not needed anymore? I would prefer following the model of 
~~~
 void picoquic_set_cookie_mode(picoquic_quic_t* quic, int cookie_mode);
~~~
That is, have a an authentication_mode parameter, with value 1 to turn it on, 0 to turn it off.

File: picoquic/quicctx.c [hunk 2]
```diff
@@ -1259,3 +1259,7 @@ void picoquic_disable_keep_alive(picoquic_cnx_t* cnx)
 {
     cnx->keep_alive_interval = 0;
 }
+
+void picoquic_enable_client_authentication(picoquic_quic_t* quic) {
+    picoquic_tls_enable_client_authentication(quic);
+}
```
Comment (huitema) [inline picoquic/quicctx.c]: See previous remark about having an authentication_mode parameter.

File: picoquic/tls_api.c [hunk 3]
```diff
@@ -1288,3 +1288,7 @@ int picoquic_create_cnxid_reset_secret(picoquic_quic_t* quic, picoquic_connectio
 
     return (ret);
 }
+
+void picoquic_tls_enable_client_authentication(picoquic_quic_t* quic) {
+    ((ptls_context_t*)quic->tls_master_ctx)->client_authentication = 1;
+}
```
Comment (huitema) [inline picoquic/tls_api.c]: Authentication mode parameter would fit naturally here.

Also, this line causes the Travis build to break:
~~~
/home/travis/build/private-octopus/picoquic/picoquic/tls_api.c: In function ‘picoquic_tls_enable_client_authentication’:
/home/travis/build/private-octopus/picoquic/picoquic/tls_api.c:1293:44: error: ‘ptls_context_t’ has no member named ‘client_authentication’
     ((ptls_context_t*)quic->tls_master_ctx)->client_authentication = 1;
~~~
Do you have a dependency on a PR in progress in Picotls?

Comment (bkchr) [reply picoquic/tls_api.c]: Yeah, I linked the pull request.

**Comment by huitema** [comment]
@bkchr: can you synch your branch with the latest fixes? I would like to merge this once we the corresponding PR is merged in Picotls, but the conflict list is a bit long -- git merge appears confused.

**Comment by bkchr** [comment]
Yeah, I will rebase the commits. The corresponding pull request for picotls requires some more work from me and then it should finally be mergable.

**Comment by huitema** [comment]
Sorry for the noise. Did not realize ghis was just an interim commit, still waiting for picotls

**Comment by bkchr** [comment]
@huitema This is still the tls client authentication pull request. I'm still waiting that my pull request is merged in picotls. Have a good flight!

**Comment by bkchr** [comment]
@huitema my changes got finally merged into picotls :) So, this pull request is mergeable from my side :)

File: picoquic/tls_api.c [hunk 4]
```diff
@@ -1371,7 +1372,14 @@ int picoquic_tlsinput_stream_zero(picoquic_cnx_t* cnx)
         case picoquic_state_server_init:
             /* Extract and install the server 0-RTT and 1-RTT key */
             picoquic_setup_0RTT_aead_contexts(cnx, 1);
-            cnx->cnx_state = picoquic_state_server_almost_ready;
+            /* If client authentication is activated, the client sends the certificates with its `Finished` package.
+               The server does not send any further packages, so, we can switch into ready state here.
+            */
```
Comment (huitema) [inline picoquic/tls_api.c]: "package" should be "packet", as in "packet switching".

**Comment by huitema** [comment]
I have just two comments. The first one is the use of the word "package" in comments to describe a unit of transmission. The standard English word for that is "packet", as in [packet switching](https://en.wikipedia.org/wiki/Packet_switching). Of course, this is purely cosmetic.

More important is the handling of a close request received in a clear text packet after the encryption keys have been negotiated. I understand why you want that, but it opens a denial of service hole for connections that have completed the handshake. It is a complex issue, so I opened bug  #165 rather than delaying this PR further, but we need to fix it.

**Pull request merged**